File: dotnet/test/Snowflake44.Tests/Snowflake44GeneratorTests.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting

Namespace Snowflake44Tests

  <TestClass>
  Public Class Snowflake44GeneratorTests

    <TestMethod()>
    Public Sub GenerateUid_CalledQuicklyAndOften_DoesNotCreateDuplicateIds()
      Dim uid As Long
      Dim previousUid As Long

      For i = 1 To 1000000

        previousUid = uid

        uid = Snowflake44.Generate()
        Assert.AreNotEqual(previousUid, uid)

      Next

    End Sub

    <TestMethod()>
    Public Sub DecodeDateTimeTest()

      Dim expectedDate1 As New Date(1973, 12, 9, 15, 33, 22)
      Dim saltedTimestamp1 As Long = Snowflake44.EncodeDateTime(expectedDate1)

      Dim expectedDate2 As Date = expectedDate1.AddMilliseconds(1)
      Dim saltedTimestamp2 As Long = Snowflake44.EncodeDateTime(expectedDate2)

      Dim actualDate1 As Date = Snowflake44.DecodeDateTime(saltedTimestamp1)
      Dim actualDate2 As Date = Snowflake44.DecodeDateTime(saltedTimestamp2)

      Assert.AreEqual(expectedDate1, actualDate1)
      Assert.AreEqual(expectedDate2, actualDate2)
      Assert.AreEqual(actualDate1.AddMilliseconds(1), actualDate2)

    End Sub

  End Class

End Namespace

File: dotnet/test/Snowflake44.TestsC/System/SmartStandards/Snowflake44GeneratorTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace System.SmartStandards {

  [TestClass]
  public class Snowflake44GeneratorTests {

    [TestMethod()]
    public void GenerateUid_CalledQuicklyAndOften_DoesNotCreateDuplicateIds() {

      long uid = 0;
      long previousUid;

      for (int i = 1; i <= 1000000; i++) {

        previousUid = uid;

        uid = Snowflake44.Generate();
        Assert.AreNotEqual(previousUid, uid);

      }

    }

    [TestMethod()]
    public void DecodeDateTimeTest() {

      DateTime expectedDate1 = new DateTime(1973, 12, 9, 15, 33, 22, DateTimeKind.Utc);
      long saltedTimestamp1 = Snowflake44.EncodeDateTime(expectedDate1);

      DateTime expectedDate2 = expectedDate1.AddMilliseconds(1d);
      long saltedTimestamp2 = Snowflake44.EncodeDateTime(expectedDate2);

      DateTime actualDate1 = Snowflake44.DecodeDateTime(saltedTimestamp1);
      DateTime actualDate2 = Snowflake44.DecodeDateTime(saltedTimestamp2);

      Assert.AreEqual(expectedDate1, actualDate1);
      Assert.AreEqual(expectedDate2, actualDate2);
      Assert.AreEqual(actualDate1.AddMilliseconds(1d), actualDate2);

    }

    [TestMethod()]
    public void GuidConversionTest() {

      const long inputUid = 2079256415714739899L;
      const string expectedGuidString = "00000000-0000-0000-1cdb-00b197a106bb";

      Guid guid = Snowflake44.ConvertToGuid(inputUid);
      Assert.AreEqual(expectedGuidString, guid.ToString());

      // check the endian correctness of the conversion
      string uidAsHex = inputUid.ToString("x16");
      string expectedLast16HexChars = expectedGuidString.Replace("-", "").Substring(16, 16);
      Assert.AreEqual(expectedLast16HexChars, uidAsHex);

      long reConvertedUid = Snowflake44.ConvertFromGuid(guid);
      Assert.AreEqual(inputUid, reConvertedUid);

    }

  }

}
